### ISSUE TITLE
please synchronise grpc with Docker

### ISSUE THREAD

**Issue by onlyjob** [docker/notary#778]
Notary and Docker use different snapshots of `google.golang.org/grpc`:
  https://github.com/docker/notary/blob/master/Godeps/Godeps.json#L759
  https://github.com/docker/docker/blob/master/hack/vendor.sh#L100

Please consider using the same revision of _grpc_ as Docker. Thanks.

**Comment by endophage** [comment]
No part of notary that depends on GRPC is vendored into docker so there appears to be no benefit to synchronizing our GRPC versions. Is there something I'm missing?

**Comment by onlyjob** [comment]
_grpc_ is shipped as re-usable package in Debian. We've updated _grpc_ to snapshot that Docker-1.11.1 needs but that broke Notary which FTBFS now because it depends on older snapshot of _grpc_... Consistency is important. ;)

**Comment by endophage** [comment]
From what you're describing here and in #782, are you saying every single package in debian needs to use the same version of a dependency? 

The Notary repository contains 3 applications. Notary CLI, Notary Server, and Notary Signer. Docker only depends on Notary CLI. Notary CLI has no dependencies on GRPC. As such I see no need to synchronize GRPC versions between the Notary and Docker repositories. This only adds unnecessary overhead for both projects.

**Comment by onlyjob** [comment]
> are you saying every single package in debian needs to use the same version of a dependency? 

Correct. Reusable libraries is the only way to maintain coherent interdependent software repository. Vendored mess quickly gets out of control or decay. Think of security support: a problem can be fixed in one place in case of reusable library or one would have to backport fix to all versions of vendored library... The latter is not viable.

All notary applications are built from single (tagged) source. If you maintain Notary components separately then you need to version them separately... Does that make sense?

For standalone Notary updating GRPC is important to fix FTBFS with newer GRPC that Docker depends upon. Sooner or later one would have to move GRPC forward otherwise Debian will stuck with outdated Notary...

**Comment by endophage** [comment]
I understand what you're saying and I've heard it many times before. I think it would really help us to get a full overview of what you're trying to do, including all the components you're trying to build packages for. It's very difficult for us to understand your use case through a handful of highly specific requests. It sounds like you're not only trying to build the Notary CLI, but the server and signer as well. 

p.s. we're not going to have any more time to respond until after DockerCon so if you don't heard anything through next week, we're not ignoring you.

p.p.s how do all the other open source projects get brought into line? Or is this why debian lags so far behind on so many versions?

**Comment by onlyjob** [comment]
> how do all the other open source projects get brought into line?

Through tagging semantically, vendoring by tags and reusing libraries. Of course some projects do not tag so we have to mock-up date-based versions...

> Or is this why debian lags so far behind on so many versions?

What do you mean? Docker was just recently uploaded and as for Notary it is in such a mess that inevitably we could not bring it up to date because of all build failures and inconsistencies that I tried to address.
It would really help if you could version/tag semantically and use only tagged releases for vendoring.
I don't see how this request is "highly specific". It is all about building software from reusable components...

**Comment by onlyjob** [comment]
> It sounds like you're not only trying to build the Notary CLI, but the server and signer as well. 

Yes, one source package produces different Notary components -- -dev package with library sources and binary packages with daemon and CLI. If you maintain those components separately shouldn't they be split into their own repositories?

**Comment by endophage** [comment]
So we weren't planning to support server and signer in any form other than docker images. That's why we don't release binaries for them on the releases page.

I'm sorry, I probably seemed like I was having a dig at Debian before. That wasn't my intent. I've used Ubuntu (which I realize isn't the same but it's the closest thing I've touched to Debian) and Arch for a littler over a decade, and the Ubuntu packages are always way, waaaay, behind. Even today it appears 16.04 is still back on docker 1.5. Debian does indeed appear to be more up to date in many cases (and not so far behind in others).

> Through tagging semantically, vendoring by tags and reusing libraries. Of course some projects do not tag so we have to mock-up date-based versions...

I guess I'm finding it difficult to suspend my disbelief that _every_ package in the debian repos at a given point in time depend on the same semantic version of, to stick with the context, grpc.

By "highly specific" I was trying to say you came to us and said "do X" and we had no context on why X was necessary. Consistency for us doesn't mean "consistency with the debian ecosystem" so we didn't have any of that context from your initial request.

**Comment by onlyjob** [comment]
Of course not every golang package is versioned but I was not making such claim. :)
"grpc" is one of those that refuses to version and it is not the only project that does not tag releases properly. Fortunately many do, notably CoreOS packages and many others.

Thanks.
